sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -233,10 +223,124 @@ class CodegenContext {
+  /**
+   * Holds the class and instance names to be generated. `OuterClass` is a placeholder standing for
+   * whichever class is generated as the outermost class and which will contain any nested
+   * sub-classes. All other classes and instance names in this list will represent private, nested
+   * sub-classes.
+   */
+  private val classes: mutable.ListBuffer[(String, String)] =
+    mutable.ListBuffer[(String, String)](("OuterClass", null))
+
+  // A map holding the current size in bytes of each class to be generated.
+  private val classSize: mutable.Map[String, Int] =
+    mutable.Map[String, Int](("OuterClass", 0))
+
+  // A map holding lists of functions belonging to their class.
+  private val classFunctions: mutable.Map[String, mutable.ListBuffer[String]] =
+    mutable.Map(("OuterClass", mutable.ListBuffer.empty[String]))
+
+  // Returns the size of the most recently added class.
+  private def currClassSize(): Int = classSize(classes.head._1)
+
+  // Returns the class name and instance name for the most recently added class.
+  private def currClass(): (String, String) = classes.head
+
+  // Adds a new class. Requires the class' name, and its instance name.
+  private def addClass(className: String, classInstance: String): Unit = {
+    classes.prepend(Tuple2(className, classInstance))
+    classSize += className -> 0
+    classFunctions += className -> mutable.ListBuffer.empty[String]
+  }
+
+  /**
+   * Adds a function to the generated class. If the code for the `OuterClass` grows too large, the
+   * function will be inlined into a new private, nested class, and a class-qualified name for the
+   * function will be returned. Otherwise, the function will be inined to the `OuterClass` the
+   * simple `funcName` will be returned.
+   *
+   * @param funcName the class-unqualified name of the function
+   * @param funcCode the body of the function
+   * @return the name of the function, qualified by class if it will be inlined to a private,
+   *         nested sub-class
+   */
+  def addNewFunction(
+    funcName: String,
+    funcCode: String,
+    inlineToOuterClass: Boolean = false): String = {
+    // The number of named constants that can exist in the class is limited by the Constant Pool
+    // limit, 65,536. We cannot know how many constants will be inserted for a class, so we use a
+    // threshold of 1600k bytes to determine when a function should be inlined to a private, nested
+    // sub-class.
+    val classInfo = if (inlineToOuterClass) {
+      ("OuterClass", "")
+    } else if (currClassSize > 1600000) {
+      val className = freshName("NestedClass")
+      val classInstance = freshName("nestedClassInstance")
+
+      addClass(className, classInstance)
+
+      className -> classInstance
+    } else {
+      currClass()
+    }
+    val name = classInfo._1
+
+    classSize.update(name, classSize(name) + funcCode.length)
+    classFunctions.update(name, classFunctions(name) += funcCode)
```

Comment:
Yeah, good point, since we're using a mutable buffer, we can update the referenced object directly even if its contained inside the map. Since `+=` explicitly returns the reference to the modified buffer, it would probably be most straightforward to use
`classFunctions(name).append(funcCode)`
since `append` has `Unit` return type, and we don't need any results from appending the code to the class's buffer of function code.